### PR TITLE
xplat CLI: Fix transitive-only case in special reports

### PR DESCRIPTION
**Pull request by drewgillies**
## Bug 

Fixes:https://github.com/NuGet/Home/issues/10767

Regression? Last working version: N/A

## Description
Long standing break for special reports (`--outdated`, `--deprecated` and now `--vulnerable`) where filtered result sets which only contain transitive dependencies are reported as empty.

Corrected output:
image

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs [hunk 1]
```diff
@@ -166,34 +152,33 @@ public async Task ExecuteCommandAsync(ListPackageArgs listPackageArgs)
...
-
-        private static bool FilterVulnerablePackages(IEnumerable<FrameworkPackages> packages)
-        {
-            FilterPackages(
-                packages,
-                ListPackageHelper.PackagesFilterForVulnerable,
-                ListPackageHelper.PackagesFilterForVulnerable);
+            switch (listPackageArgs.ReportType)
+            {
+                case ReportType.Default: break; // We do no filtering in this case
```
Comment (dominoFire) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs]:
```suggestion
                case ReportType.Default: break; // no filtering in this case
```

**Comment by drewgillies** [comment]
Note that I've pulled the functional test `DotnetListPackage_ProjectReference_Succeeds` - this test was failing because it required a project reference to appear as a transitive package. This strikes me as counter to the purpose of the report, which is _list packages_. Open to other opinions, of course.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs [hunk 2]
```diff
@@ -449,7 +449,7 @@ internal static bool IsPackageReferenceProject(Project project)
                     }
                     // If no matching packages were found, then the package is transitive,
                     // and include-transitive must be used to add the package
-                    else if (transitive)
+                    else if (transitive && (library.Type == "package")) // be sure to exclude "project" references here as these are irrelevant
```
Comment (dominoFire) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: You can remove the parenthesis

```suggestion
                    else if (transitive && library.Type == "package") // be sure to exclude "project" references here as these are irrelevant

```

**Review by dominoFire — COMMENTED** [review]
Partial review

**Comment by nkolev92** [comment]
> Note that I've pulled the functional test DotnetListPackage_ProjectReference_Succeeds - this test was failing because it required a project reference to appear as a transitive package. This strikes me as counter to the purpose of the report, which is list packages. Open to other opinions, of course.

Packages can be pulled through projects, maybe the idea was to display that to allow the dependency graph to be traced?
Packages and projects are treated equivalently. 

It's not called out explicitly in the design, https://github.com/NuGet/Home/wiki/dotnet-list-package. 

I think if we're removing the test we should remove the functionality as well.

**Comment by dominoFire** [comment]
> > Note that I've pulled the functional test DotnetListPackage_ProjectReference_Succeeds - this test was failing because it required a project reference to appear as a transitive package. This strikes me as counter to the purpose of the report, which is list packages. Open to other opinions, of course.
> 
> Packages can be pulled through projects, maybe the idea was to display that to allow the dependency graph to be traced?
> Packages and projects are treated equivalently.
> 
> It's not called out explicitly in the design, https://github.com/NuGet/Home/wiki/dotnet-list-package.
> 
> I think if we're removing the test we should remove the functionality as well.

@nkolev92 @drewgillies  Can we address this in another issue and keep the current test ?: https://github.com/NuGet/Home/issues/10817

**Comment by nkolev92** [comment]
> @nkolev92 @drewgillies Can we address this in another issue and keep the current test ?: NuGet/Home#10817

For sure!
Removing functionality should go through a separate issue and design. Personally I'm not sure what the best experience is for this. I think we need to talk to some customers.

**Comment by dominoFire** [comment]
@drewgillies Please review new changes. Thanks!

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx [hunk 3]
```diff
@@ -789,4 +789,8 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
   <data name="TrustCommandOwners" xml:space="preserve">
     <value>Semi-colon separated list of trusted owners to further restrict the trust of a repository.</value>
   </data>
+  <data name="Exception_TransitivePackagesWithoutTransitiveFlag" xml:space="preserve">
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx]: Is this used anywhere?

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs [hunk 4]
```diff
@@ -323,10 +323,12 @@ internal static bool IsPackageReferenceProject(Project project)
...
         /// <param name="assetsFile">Assets file for all targets and libraries</param>
-        /// <param name="transitive">Include transitive packages in the result</param>
-        /// <returns></returns>
+        /// <param name="transitive">Include transitive packages/projects in the result</param>
+        /// <param name="includeProjects">Include project references in top-level and transitive package lists</param>
+        /// <returns>FrameworkPackages collection with top-level and transitive package/project
+        /// references for each framework, or null on error</returns>
         internal IEnumerable<FrameworkPackages> GetResolvedVersions(
-            string projectPath, IEnumerable<string> userInputFrameworks, LockFile assetsFile, bool transitive)
+            string projectPath, IEnumerable<string> userInputFrameworks, LockFile assetsFile, bool transitive, bool includeProjects)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:331]: The projects are *only* listed in `dotnet list package`? 

outdated, deprecated and vulnerable don't list them, is that correct?

**Review by nkolev92 — COMMENTED** [review]
Looks good.
Minor questions.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx (hunk 3 above)
Comment (dominoFire) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx]: No, will remove.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs (hunk 4 above)
Comment (dominoFire) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:331]: > The projects are _only_ listed in `dotnet list package`?

That's correct, the default report.

> outdated, deprecated and vulnerable don't list them, is that correct?

Yes.

**Review by nkolev92 — APPROVED** [review]

**Pull request merged**